### PR TITLE
swb: strictly check write barrier bit

### PR DESCRIPTION
**Pull request by leirocks**
strictly check where write barrier bit is set for none write barrier pages, in order to check the unnecessary annotation

### CODE REVIEW

File: lib/Common/ConfigFlagsList.h [hunk 1]
```diff
@@ -1492,6 +1492,9 @@ FLAGNR(Boolean, CFG, "Force enable CFG on jshost. version in the jshost's manife
 
 FLAGR(Number, JITServerIdleTimeout, "Idle timeout in seconds to do the cleanup in JIT server", 10)
 FLAGR(Number, JITServerMaxInactivePageAllocatorCount, "Max inactive page allocators to keep before schedule a cleanup", 10)
+
+FLAGNR(Boolean, StrictWriteBarrierCheck, "Check write barrier setting on none write barrier pages", false)
```
Comment (jianchun) [inline lib/Common/ConfigFlagsList.h]: Do we want this flag available on "Test" build? If yes, change caller RecyclerWriteBarrierManager to use it in "Test" build (DBG -> ENABLE_DEBUG_CONFIG_OPTIONS).

File: lib/Common/Memory/RecyclerWriteBarrierManager.cpp [hunk 2]
```diff
@@ -30,6 +33,7 @@ BYTE* RecyclerWriteBarrierManager::cardTable = RecyclerWriteBarrierManager::x64C
 #else
 // Each byte in the card table covers 4096 bytes so the range covered by the table is 4GB
 BYTE RecyclerWriteBarrierManager::cardTable[1 * 1024 * 1024];
+bool dummy = RecyclerWriteBarrierManager::Initialize();
```
Comment (digitalinfinity) [inline lib/Common/Memory/RecyclerWriteBarrierManager.cpp]: Why is this needed?

**Comment by digitalinfinity** [comment]
@leirocks can you provide more details on the purpose of this change? What exactly does "strict barrier check" mean?

File: lib/Common/Memory/RecyclerWriteBarrierManager.cpp (hunk 2 above)
Comment (jianchun) [inline lib/Common/Memory/RecyclerWriteBarrierManager.cpp]: I was looking at this too; I read that it was to initialize the "g_..." function for the other "CopyArray" code to verify "no barrier needed" when the other no barrier variants were used. But this looks heavy weighted and incur penalty for general case -- looks the other code always call this "g_..." function no matter what. I think you should be able to (1) only do the verification in debug/test builds, and (2) forward declare a global function instead of creating this global variable.

---
In reply to: [86642266](https://github.com/Microsoft/ChakraCore/pull/1920#discussion_r86642266) [](ancestors = 86642266)

File: lib/Common/Memory/Recycler.h [hunk 3]
```diff
@@ -2353,6 +2385,7 @@ class ListTypeAllocatorFunc<Recycler, isLeaf>
 // Dummy class to choose the allocation function
 class RecyclerLeafAllocator;
 class RecyclerNonLeafAllocator;
+class RecyclerWriteBarrierAllocator;
```
Comment (jianchun) [inline lib/Common/Memory/Recycler.h]: >RecyclerWriteBarrierAllocator [](start = 6, length = 29)

Could you explain why this new allocator type is needed and where it is used? This makes choosing a Recycler allocator more complex.

File: lib/Common/Memory/RecyclerPointers.h [hunk 4]
```diff
@@ -58,8 +59,23 @@ struct AllocatorWriteBarrierPolicy
...
+
 template <class T>
-struct AllocatorWriteBarrierPolicy<RecyclerNonLeafAllocator, T> { typedef _write_barrier_policy Policy; };
+struct AllocatorWriteBarrierPolicy<RecyclerNonLeafAllocator, T> 
+{
+#if GLOBAL_FORCE_USE_WRITE_BARRIER
+    typedef _write_barrier_policy Policy;
+#else
+    typedef _no_write_barrier_policy Policy; 
+#endif
```
Comment (jianchun) [inline lib/Common/Memory/RecyclerPointers.h]: As offline conversation, a cleaner way might be introducing a 3rd policy variable -- default global policy, which is on by default xplat; off by default Windows,on if globally force. The effective policy would be determined by: allocator type, data type, default global policy (I could work on that after my current annotation batch.)

**Comment by leirocks** [comment]
JC  wants to use the tool to annotate all possible fields, but some of them are not necessary, because some of the struct is never recycler allocated with write barrier. this change will Assert on those address which not allocated with barrier but we are setting barrier bit, so with this we can find out which struct does not need annotate.  this can save a lot of work to annotate the structures instead of understanding usage of each structure and field.

this can also tell us which structure has annotation allocated both with or without barrier , we may specialize it to have both annotated version and non-annotated version



---
In reply to: [258561828](https://github.com/Microsoft/ChakraCore/pull/1920#issuecomment-258561828) [](ancestors = 258561828)

File: lib/Common/Memory/Recycler.h (hunk 3 above)
Comment (leirocks) [inline lib/Common/Memory/Recycler.h]: because not all NonLeaf are allocated with software write barrier, on windows, swb is only enabled in a few class

---
In reply to: [86644009](https://github.com/Microsoft/ChakraCore/pull/1920#discussion_r86644009) [](ancestors = 86644009)

File: lib/Common/Memory/RecyclerWriteBarrierManager.cpp (hunk 2 above)
Comment (leirocks) [inline lib/Common/Memory/RecyclerWriteBarrierManager.cpp]: because of the initialization order, before the card table is allocated, we are already allocating dictionary etc with heap allocator. and in "struct _ArrayWriteBarrier" RecyclerWriteBarrierManager is not availble yet, otherwise we need to change the include and initialization order
the g_ function should only be called when the flag is enabled, I'll fix that part.

---
In reply to: [86643214](https://github.com/Microsoft/ChakraCore/pull/1920#discussion_r86643214) [](ancestors = 86643214,86642266)

File: lib/Common/ConfigFlagsList.h (hunk 1 above)
Comment (leirocks) [inline lib/Common/ConfigFlagsList.h]: maybe yes, not sure if there are many difference on the data structures. let me enable it on test build

---
In reply to: [86637769](https://github.com/Microsoft/ChakraCore/pull/1920#discussion_r86637769) [](ancestors = 86637769)

File: lib/Common/Memory/RecyclerPointers.h [hunk 5]
```diff
@@ -109,23 +135,33 @@ void WriteBarrier(T * address, size_t count)
...
+        return CopyArray<Allocator, T, PolicyType>((T*)dst, dstCount, (const T*)src, srcCount);
+    }
+    static void CopyArray(WriteBarrierPtr<T>& dst, size_t dstCount, const WriteBarrierPtr<T>& src, size_t srcCount)
+    {
+        return CopyArray<Allocator, T, PolicyType>((T*)dst, dstCount, (const T*)src, srcCount);
+    }
+};
 
+template <class T, class PolicyType>
+struct WriteBarrierUtil<Recycler, T, PolicyType>
```
Comment (leirocks) [inline lib/Common/Memory/RecyclerPointers.h]: because function template can't be partial specialized

---
In reply to: [86643736](https://github.com/Microsoft/ChakraCore/pull/1920#discussion_r86643736) [](ancestors = 86643736)

File: lib/Common/Memory/RecyclerWriteBarrierManager.h [hunk 6]
```diff
@@ -124,6 +145,32 @@ class RecyclerWriteBarrierManager
 public:
     static void WriteBarrier(void * address);
     static void WriteBarrier(void * address, size_t bytes);
+    static void ToggleBarrier(void * address, size_t bytes, bool enable);
+    static BOOL IsBarrierAddress(void * address);
+    static BOOL IsBarrierAddress(uintptr_t index);
+    static void VerifyIsBarrierAddress(void * address, size_t bytes);
+    static void VerifyIsNotBarrierAddress(void * address, size_t bytes);
+    static void VerifyIsBarrierAddress(void * address);
+    static bool Initialize();
```
Comment (leirocks) [inline lib/Common/Memory/RecyclerWriteBarrierManager.h]: the linker should drop them, but yes, should make them only applicable for test build

---
In reply to: [86644302](https://github.com/Microsoft/ChakraCore/pull/1920#discussion_r86644302) [](ancestors = 86644302)

File: lib/Common/Exceptions/ReportError.cpp [hunk 7]
```diff
@@ -19,6 +19,9 @@ void ReportFatalException(
     if (IsDebuggerPresent())
     {
         DebugBreak();
+#if DBG
+        return;
+#endif
```
Comment (jianchun) [inline lib/Common/Exceptions/ReportError.cpp]: Curious what this change is for? Is it to enable continue execution in debugger after an assertion?

Comment (leirocks) [inline lib/Common/Exceptions/ReportError.cpp]: that's just easier to debug, forgot to remove this after rebase/merge with my local change

File: lib/Common/Memory/PageAllocator.cpp [hunk 8]
```diff
@@ -56,7 +57,20 @@ SegmentBase<T>::~SegmentBase()
...
+            {
+                RecyclerWriteBarrierManager::OnSegmentFree(this->address, this->segmentPageCount);
+                RecyclerWriteBarrierManager::ToggleBarrier(this->address, this->segmentPageCount * AutoSystemInfo::PageSize, false);
+            }
+        }
+        else
+#endif
+        {
+            RecyclerWriteBarrierManager::OnSegmentFree(this->address, this->segmentPageCount);
+        }
```
Comment (jianchun) [inline lib/Common/Memory/PageAllocator.cpp]: Looks this fragment can be simplified a bit. Always call "OnSegmentFree". After that, do "#if...if... call ToggleBarrier". Avoids duplicate code "OnSegmentFree".

File: lib/Common/Memory/PageAllocator.cpp [hunk 9]
```diff
@@ -138,17 +152,39 @@ SegmentBase<T>::Initialize(DWORD allocFlags, bool excludeGuardPages)
...
+            // only commit card table for write barrier pages for strict check
+            // we can do this in free build if all write barrier annotated struct only allocate with write barrier pages
+            registerBarrierResult = RecyclerWriteBarrierManager::OnSegmentAlloc(this->address, this->segmentPageCount);
+        }
+    }
+    else
+#endif
+    {
+        registerBarrierResult = RecyclerWriteBarrierManager::OnSegmentAlloc(this->address, this->segmentPageCount);
+    }
```
Comment (jianchun) [inline lib/Common/Memory/PageAllocator.cpp]: Why "if" and "else" body looks the same?

File: lib/Common/Memory/PageAllocator.cpp [hunk 10]
```diff
@@ -138,17 +152,39 @@ SegmentBase<T>::Initialize(DWORD allocFlags, bool excludeGuardPages)
...
         GetAllocator()->GetVirtualAllocator()->Free(originalAddress, GetPageCount() * AutoSystemInfo::PageSize, MEM_RELEASE, this->GetAllocator()->processHandle);
         this->GetAllocator()->ReportFailure(GetPageCount() * AutoSystemInfo::PageSize);
         this->address = nullptr;
         return false;
     }
-    else
+#endif
+
+    this->isWriteBarrierAllowed = true;
+#if DBG
```
Comment (jianchun) [inline lib/Common/Memory/PageAllocator.cpp]: >DBG [](start = 4, length = 3)

Should this be ENABLE_DEBUG_CONFIG_OPTIONS?

File: lib/Common/Memory/Recycler.h [hunk 11]
```diff
@@ -1355,6 +1355,7 @@ class Recycler
     DEFINE_RECYCLER_ALLOC(AllocTrackedWithBarrier, ClientTrackableObjectWithBarrierBits);
     DEFINE_RECYCLER_ALLOC(AllocFinalizedClientTrackedWithBarrier, ClientFinalizableObjectWithBarrierBits);
 #endif
+
```
Comment (jianchun) [inline lib/Common/Memory/Recycler.h]: nit: spaces

File: lib/Common/Memory/RecyclerPointers.h [hunk 12]
```diff
@@ -220,6 +235,9 @@ class NoWriteBarrierField
     // Setters
     NoWriteBarrierField& operator=(T const& value)
     {
+#if defined(RECYCLER_WRITE_BARRIER)
+        RecyclerWriteBarrierManager::VerifyIsBarrierAddress(this);
```
Comment (jianchun) [inline lib/Common/Memory/RecyclerPointers.h]: Should this be VerifyIs"Not"? Should this be under #if defined(RECYCLER_WRITE_BARRIER) && ENABLE_DEBUG_CONFIG_OPTIONS, && commandline options?

File: lib/Common/Memory/RecyclerWriteBarrierManager.cpp [hunk 13]
```diff
@@ -20,7 +20,10 @@
 #pragma init_seg(".CRT$XCAU")
 
 #ifdef RECYCLER_WRITE_BARRIER
-
+namespace Memory 
+{
+    FN_VerifyIsNotBarrierAddress* g_verifyIsNotBarrierAddress = nullptr;
+}
```
Comment (jianchun) [inline lib/Common/Memory/RecyclerWriteBarrierManager.cpp]: Wrap under #if defined(RECYCLER_WRITE_BARRIER) && ENABLE_DEBUG_CONFIG_OPTIONS

File: lib/Common/Memory/PageAllocator.cpp (hunk 8 above)
Comment (leirocks) [inline lib/Common/Memory/PageAllocator.cpp]: yes, since OnSegmentFree do nothing. otherwise if it decommit card table we might still need current pattern, need toggle barrier first though

---
In reply to: [87499422](https://github.com/Microsoft/ChakraCore/pull/1920#discussion_r87499422) [](ancestors = 87499422)

File: lib/Common/Memory/RecyclerWriteBarrierManager.cpp [hunk 14]
```diff
@@ -218,6 +226,10 @@ X64WriteBarrierCardTableManager::Initialize()
 {
     AutoCriticalSection critSec(&_cardTableInitCriticalSection);
 
+#if ENABLE_DEBUG_CONFIG_OPTIONS
+    RecyclerWriteBarrierManager::Initialize();
+#endif
+
```
Comment (jianchun) [inline lib/Common/Memory/RecyclerWriteBarrierManager.cpp]: Is there any potential concurrent concern? If so maybe move this after _cardTable initialized?

File: lib/Common/Memory/RecyclerWriteBarrierManager.h [hunk 15]
```diff
@@ -63,12 +76,20 @@ class X64WriteBarrierCardTableManager
...
     BYTE * GetAddressOfCardTable() { return _cardTable; }
 
-#if DBG
+#if ENABLE_DEBUG_CONFIG_OPTIONS
     void AssertWriteToAddress(_In_ void* address)
     {
         Assert(_cardTable);
         Assert(committedSections.Test(GetSectionIndex(address)));
     }
+    BOOLEAN IsCardTableCommited(_In_ uintptr_t index)
```
Comment (jianchun) [inline lib/Common/Memory/RecyclerWriteBarrierManager.h]: >BOOLEAN [](start = 4, length = 7)

Any reason not to use "bool"? Same applies to others in this file.

Comment (leirocks) [inline lib/Common/Memory/RecyclerWriteBarrierManager.h]: need to cast then, prefast warns perf impact, it's debug code though, let me change it

---
In reply to: [87501538](https://github.com/Microsoft/ChakraCore/pull/1920#discussion_r87501538) [](ancestors = 87501538)

**Comment by jianchun** [comment]
:shipit:
